contracts/implementations/nodes/VirtualDevice.sol

```diff
@@ -0,0 +1,234 @@
+    /// @dev Only an admin can set the address
+    /// @param addr The address of the proxy
+    function setVirtualDeviceIdProxyAddress(address addr)
+        external
+        onlyRole(DEFAULT_ADMIN_ROLE)
+    {
+        require(addr != address(0), "Non zero address");
+        VirtualDeviceStorage.getStorage().idProxyAddress = addr;
+
+        emit VirtualDeviceIdProxySet(addr);
+    }
+
+    /// @notice Adds an attribute to the whielist
+    /// @dev Only an admin can add a new attribute
+    /// @param attribute The attribute to be added
+    function addVirtualDeviceAttribute(string calldata attribute)
+        external
+        onlyRole(DEFAULT_ADMIN_ROLE)
+    {
+        require(
+            AttributeSet.add(
+                VirtualDeviceStorage.getStorage().whitelistedAttributes,
+                attribute
+            ),
+            "Attribute already exists"
+        );
+
+        emit VirtualDeviceAttributeAdded(attribute);
+    }
+
+    // ***** Interaction with nodes *****//
+
+    /**
+     * @notice Mints a virtual device and pair it with a vehicle
+     * @dev Caller must have the admin role
+     * @param data input data with the following fields:
+     *  integrationNode -> Parent integration node id
+     *  vehicleNode -> Vehicle node id
+     *  virtualDeviceSig -> Virtual Device's signature hash
+     *  vehicleOwnerSig -> Vehicle owner signature hash
+     *  virtualDeviceAddr -> Address associated with the virtual device
+     *  attrInfoPairs -> List of attribute-info pairs to be added
+     */
+    function mintVirtualDeviceSign(MintVirtualDeviceInput calldata data)
+        external
+        onlyRole(DEFAULT_ADMIN_ROLE)
+    {
+        NodesStorage.Storage storage ns = NodesStorage.getStorage();
+        MapperStorage.Storage storage ms = MapperStorage.getStorage();
+        VirtualDeviceStorage.Storage storage vds = VirtualDeviceStorage
+            .getStorage();
+
+        address vehicleIdProxyAddress = VehicleStorage
+            .getStorage()
+            .idProxyAddress;
+        address virtualDeviceIdProxyAddress = vds.idProxyAddress;
+
+        require(
+            INFT(IntegrationStorage.getStorage().idProxyAddress).exists(
+                data.integrationNode
+            ),
+            "Invalid parent node"
+        );
+        require(
+            INFT(vehicleIdProxyAddress).exists(data.vehicleNode),
+            "Invalid vehicle node"
+        );
+        require(
+            vds.deviceAddressToNodeId[data.virtualDeviceAddr] == 0,
+            "Device address already registered"
+        );
+        require(
+            ms.vehicleVirtualDeviceLinks[vehicleIdProxyAddress][
+                virtualDeviceIdProxyAddress
+            ][data.vehicleNode] == 0,
+            "Vehicle already paired"
+        );
+
+        address owner = INFT(vehicleIdProxyAddress).ownerOf(data.vehicleNode);
+        uint256 newTokenId = INFT(virtualDeviceIdProxyAddress).safeMint(owner);
```

Comment:
Why do we mint before signature verification? Just curious.

Reply:
Bc I forgot to move it down after some changes.